app/controllers/spree/orders_controller_decorator.rb

```diff
@@ -224,22 +224,27 @@ def wrap_json_infinity(n)
     n == Float::INFINITY ? 2147483647 : n
   end
 
+  # If a specific order is mentioned, check that it is COMPLETE, that changes
+  # are allowed and that the user has access. Return nil if not.
   def order_to_update
-    order = Spree::Order.complete.find_by_number(params[:id])
-    return order if order.andand.changes_allowed? && can?(:update, order)
-    current_order
+    return @order_to_update if defined? @order_to_update
+    return @order_to_update = current_order unless params[:id]
+    @order_to_update = begin
+      order = Spree::Order.complete.find_by_number(params[:id])
+      order = nil unless order.andand.changes_allowed? && can?(:update, order)
+      order
+    end
```

Comment:
I can, thought it was on the borderline between readability and excessive breakdown of the method.